### PR TITLE
Add code comments and minor polish

### PR DESCRIPTION
**Pull request by deepu105**

### CODE REVIEW

File: dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/DynamicAuthViewModel.kt [hunk 1]
```diff
@@ -27,17 +27,7 @@ import com.okta.idx.android.dynamic.SocialRedirectCoordinator
...
-import com.okta.idx.kotlin.dto.IdxIdpCapability
-import com.okta.idx.kotlin.dto.IdxNumberChallengeCapability
-import com.okta.idx.kotlin.dto.IdxPollAuthenticatorCapability
-import com.okta.idx.kotlin.dto.IdxPollRemediationCapability
-import com.okta.idx.kotlin.dto.IdxRecoverCapability
-import com.okta.idx.kotlin.dto.IdxRemediation
-import com.okta.idx.kotlin.dto.IdxResendCapability
-import com.okta.idx.kotlin.dto.IdxResponse
-import com.okta.idx.kotlin.dto.IdxTotpCapability
+import com.okta.idx.kotlin.dto.*
```
Comment (JayNewstrom) [inline dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/DynamicAuthViewModel.kt:30]: Please turn star imports off, and include all imports individually.

File: dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/DynamicAuthViewModel.kt [hunk 2]
```diff
@@ -126,19 +122,23 @@ internal class DynamicAuthViewModel(private val recoveryToken: String) : ViewMod
...
     private suspend fun handleResponse(response: IdxResponse) {
+        // If a response is successful, immediately exchange it for a token and set it on LoggedInUserView
         if (response.isLoginSuccessful) {
             when (val exchangeCodesResult =
                 client?.exchangeInteractionCodeForTokens(response.remediations[IdxRemediation.Type.ISSUE]!!)) {
                 is IdxClientResult.Error -> {
                     _state.value = DynamicAuthState.Error("Failed to call resume")
                 }
                 is IdxClientResult.Success -> {
+                    cancelPolling()
```
Comment (JayNewstrom) [inline dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/DynamicAuthViewModel.kt:133]: While this doesn't hurt, it's not really needed, it'll be cancelled automatically once we navigate away from this page.

File: dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/DynamicAuthViewModel.kt [hunk 3]
```diff
@@ -338,6 +368,9 @@ internal class DynamicAuthViewModel(private val recoveryToken: String) : ViewMod
         proceed(remediation)
     }
 
+    /**
+     * Proceed to the next step in IDX flow using the current remediation
```
Comment (JayNewstrom) [inline dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/DynamicAuthViewModel.kt:372]:
```suggestion
     * Proceed to the next step in the IDX flow using the specified remediation.
```

File: dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/IdxClientConfigurationProvider.kt [hunk 4]
```diff
@@ -19,6 +19,9 @@ import com.okta.idx.android.dynamic.BuildConfig
 import com.okta.idx.kotlin.client.IdxClientConfiguration
 import okhttp3.HttpUrl.Companion.toHttpUrl
 
+/**
+ * Provides OKTA org configurations for IDX client
```
Comment (JayNewstrom) [inline dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/IdxClientConfigurationProvider.kt:23]:
```suggestion
 * Provides Okta organization configuration for IDX client.
```

File: dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/DynamicAuthField.kt [hunk 5]
```diff
@@ -22,7 +22,14 @@ import androidx.lifecycle.MutableLiveData
 import com.okta.idx.android.util.emitValidation
 import com.okta.idx.kotlin.dto.IdxRemediation
 
+/**
```
Comment (JayNewstrom) [inline dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/DynamicAuthField.kt:25]: The documentation in this class is confusing to me, as it appears to talk in 2 different levels of abstraction. Can we revert the changes here, and work on those in a follow up?

File: dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/DynamicAuthFragment.kt [hunk 6]
```diff
@@ -15,43 +15,30 @@
...
-import com.okta.idx.android.dynamic.databinding.ErrorFieldBinding
-import com.okta.idx.android.dynamic.databinding.FormActionPrimaryBinding
-import com.okta.idx.android.dynamic.databinding.FormCheckBoxBinding
-import com.okta.idx.android.dynamic.databinding.FormImageBinding
-import com.okta.idx.android.dynamic.databinding.FormOptionBinding
-import com.okta.idx.android.dynamic.databinding.FormOptionsBinding
-import com.okta.idx.android.dynamic.databinding.FormTextBinding
-import com.okta.idx.android.dynamic.databinding.FragmentDynamicAuthBinding
-import com.okta.idx.android.dynamic.databinding.LoadingBinding
+import com.okta.idx.android.dynamic.databinding.*
```
Comment (JayNewstrom) [inline dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/DynamicAuthFragment.kt:37]: Please disable star imports, and import everything individually.

File: dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/DynamicAuthFragment.kt [hunk 7]
```diff
@@ -72,6 +59,7 @@ internal class DynamicAuthFragment : BaseFragment<FragmentDynamicAuthBinding>(
             when (state) {
                 is DynamicAuthState.Form -> {
                     addMessageViews(state.messages)
+                    // if there are dynamic fields remove current view, iterate through fields and render them
```
Comment (JayNewstrom) [inline dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/DynamicAuthFragment.kt:62]: Please make all comments sentences, justification here: https://nedbatchelder.com/blog/201401/comments_should_be_sentences.html

File: dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/DynamicAuthViewModel.kt [hunk 8]
```diff
@@ -324,6 +351,9 @@ internal class DynamicAuthViewModel(private val recoveryToken: String) : ViewMod
         pollingJob = null
     }
 
+    /**
+     * Start aan activity from redirect or proceed to the next step in IDX flow using the current remediation
```
Comment (JayNewstrom) [inline dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/DynamicAuthViewModel.kt:355]: This is missing some clarity. "Start an activity from redirect" doesn't give me the context I need to understand this.

**Review by JayNewstrom — CHANGES_REQUESTED** [review]

**Comment by deepu105** [comment]
@JayNewstrom I have made most of the changes could you help to improve the comments on the DynamicAuthField :pray:

File: dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/DynamicAuthViewModel.kt [hunk 9]
```diff
@@ -163,24 +175,34 @@ internal class DynamicAuthViewModel(private val recoveryToken: String) : ViewMod
...
+        // Check for messages, such as entering an incorrect code or auth error.
         val messages = mutableListOf<String>()
         for (message in response.messages) {
             messages += message.message
         }
         _state.value = DynamicAuthState.Form(response, fields, messages)
     }
 
+    /**
+     * Extract a label from a number challenge in IdxRemediation.authenticators.
```
Comment (JayNewstrom) [inline dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/DynamicAuthViewModel.kt:187]: This comment seems to not match what the method is doing.

File: dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/DynamicAuthViewModel.kt [hunk 10]
```diff
@@ -190,15 +212,20 @@ internal class DynamicAuthViewModel(private val recoveryToken: String) : ViewMod
         return DynamicAuthField.Image(label, bitmap, capability.sharedSecret)
     }
 
+    /**
+     * Extract text fields and radio groups from IdxRemediation.form.visibleFields.
```
Comment (JayNewstrom) [inline dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/DynamicAuthViewModel.kt:216]: This is incomplete, this method does much more.

File: dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/DynamicAuthViewModel.kt [hunk 11]
```diff
@@ -234,6 +263,9 @@ internal class DynamicAuthViewModel(private val recoveryToken: String) : ViewMod
         }
     }
 
+    /**
+     * Create Resend action for IdxRemediation when capabilities are empty.
```
Comment (JayNewstrom) [inline dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/DynamicAuthViewModel.kt:267]: This comment is incorrect.

File: dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/DynamicAuthViewModel.kt [hunk 12]
```diff
@@ -271,13 +306,19 @@ internal class DynamicAuthViewModel(private val recoveryToken: String) : ViewMod
         })
     }
 
+    /**
+     * Create Recover action for IdxResponse when capabilities are empty.
```
Comment (JayNewstrom) [inline dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/DynamicAuthViewModel.kt:310]: This comment is incorrect.

File: dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/DynamicAuthField.kt [hunk 13]
```diff
@@ -22,7 +22,14 @@ import androidx.lifecycle.MutableLiveData
...
+ * Data model classes to hold dynamic fields sent in remediation by IDX SDK response.
```
Comment (JayNewstrom) [inline dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/DynamicAuthField.kt:26]:
```suggestion
 * Data model classes representing `IdxRemediation` as a View Model. 
```

File: dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/DynamicAuthField.kt [hunk 14]
```diff
@@ -22,7 +22,14 @@ import androidx.lifecycle.MutableLiveData
...
+ * Fields sent are: Text fields, password fields, QR Codes, checkbox, options, labels and action buttons).
```
Comment (JayNewstrom) [inline dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/DynamicAuthField.kt:27]:
```suggestion
```

File: dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/DynamicAuthField.kt [hunk 15]
```diff
@@ -22,7 +22,14 @@ import androidx.lifecycle.MutableLiveData
...
+ */
 sealed class DynamicAuthField {
+    /**
+     * Text class holds text and password fields. Supports inline validation.
```
Comment (JayNewstrom) [inline dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/DynamicAuthField.kt:31]:
```suggestion
     * `DynamicAuthField.Text` are displayed as a `TextInputLayout`, and represents a `IdxRemediation.Form.Field.type` `string` fields.
```

File: dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/DynamicAuthField.kt [hunk 16]
```diff
@@ -40,14 +47,17 @@ sealed class DynamicAuthField {
...
+                _errorsLiveData.emitValidation { value.isNotEmpty() }
             } else {
-                return true
+                true
             }
         }
     }
 
+    /**
+     * Checkbox class holds boolean fields.
```
Comment (JayNewstrom) [inline dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/DynamicAuthField.kt:59]:
```suggestion
     * `DynamicAuthField.Checkbox` is displayed as a `Checkbox`, and represents a `IdxRemediation.Form.Field.type` `boolean` fields.
```

File: dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/DynamicAuthField.kt [hunk 17]
```diff
@@ -59,6 +69,11 @@ sealed class DynamicAuthField {
             }
     }
 
+    /**
+     * Options class holds nested radio buttons, usually used for authenticator selection and so on.
```
Comment (JayNewstrom) [inline dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/DynamicAuthField.kt:73]:
```suggestion
     * `DynamicAuthField.Options` is displayed as a `RadioButton`, and represents a `IdxRemediation.Form.Field` with `options` fields.
```

File: dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/DynamicAuthField.kt [hunk 18]
```diff
@@ -86,32 +101,41 @@ sealed class DynamicAuthField {
...
+    /**
+     * Action class holds actions sent by remediations and are usually rendered as buttons.
+     */
     data class Action(
         val label: String,
         val onClick: (context: Context) -> Unit
     ) : DynamicAuthField()
 
+    /**
+     * Image holds bitmap QR code data sent for an authenticator enrollment in remediation.
```
Comment (JayNewstrom) [inline dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/DynamicAuthField.kt:128]:
```suggestion
     * `DynamicAuthField.Image` is displayed as an `ImageView`, and represents an `IdxRemediation.authenticators` capability of `IdxTotpCapability`.
```

File: dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/DynamicAuthField.kt [hunk 19]
```diff
@@ -86,32 +101,41 @@ sealed class DynamicAuthField {
...
+                _errorsLiveData.emitValidation { option != null } && nestedFieldsAreValid
             } else {
-                return true
+                true
             }
         }
     }
 
+    /**
+     * Action class holds actions sent by remediations and are usually rendered as buttons.
```
Comment (JayNewstrom) [inline dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/DynamicAuthField.kt:120]:
```suggestion
     * `DynamicAuthField.Action` is displayed as a `Button`, and typically represents a call submitting an `IdxRemediation` to `IdxClient.proceed`.
```

**Review by JayNewstrom — COMMENTED** [review]
The content doesn't have to stay exactly as I updated, but the level of abstraction is consistent throughout in the way I changed it.

File: dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/DynamicAuthViewModel.kt [hunk 20]
```diff
@@ -126,19 +132,23 @@ internal class DynamicAuthViewModel(private val recoveryToken: String) : ViewMod
     }
 
     private suspend fun handleResponse(response: IdxResponse) {
+        // If a response is successful, immediately exchange it for a token and set it on LoggedInUserView.
```
[the remark this answers is not in the document]
Comment (deepu105) [reply dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/DynamicAuthViewModel.kt:135]: removed

File: dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/DynamicAuthViewModel.kt (hunk 10 above)
Comment (deepu105) [reply dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/DynamicAuthViewModel.kt:216]: updated, see if ok, else please suggest

File: dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/DynamicAuthViewModel.kt (hunk 11 above)
Comment (deepu105) [reply dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/DynamicAuthViewModel.kt:267]: corrected, see if ok

File: dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/DynamicAuthViewModel.kt (hunk 12 above)
Comment (deepu105) [reply dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/DynamicAuthViewModel.kt:310]: updated

File: dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/DynamicAuthFragment.kt [hunk 21]
```diff
@@ -121,14 +121,19 @@ internal class DynamicAuthFragment : BaseFragment<FragmentDynamicAuthBinding>(
         parent.addView(binding.root)
     }
 
+    /**
+     * Render `DynamicAuthField` models dynamically on the given view.
```
Comment (JayNewstrom) [inline dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/DynamicAuthFragment.kt:125]:
```suggestion
     * Create a `View` to render the `DynamicAuthField`.
```

File: dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/DynamicAuthViewModel.kt [hunk 22]
```diff
@@ -153,8 +163,10 @@ internal class DynamicAuthViewModel(private val recoveryToken: String) : ViewMod
             }
             fields += remediation.asDynamicAuthFieldResendAction()
             fields += remediation.asDynamicAuthFieldActions()
+            // Start polling current remediation. Required for asynchronous actions like using an email magic link to sign in.
```
Comment (JayNewstrom) [inline dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/DynamicAuthViewModel.kt:166]:
```suggestion
            // Start polling current remediation (if applicable). Required for asynchronous actions like using an email magic link to sign in.
```

File: dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/DynamicAuthViewModel.kt [hunk 23]
```diff
@@ -287,24 +328,20 @@ internal class DynamicAuthViewModel(private val recoveryToken: String) : ViewMod
         })
     }
 
+    /**
+     * Start polling on a remediation for asynchronous actions like clicking on an email magic link or okta verify.
```
Comment (JayNewstrom) [inline dynamic-app/src/main/java/com/okta/idx/android/dynamic/auth/DynamicAuthViewModel.kt:332]:
```suggestion
     * Start polling on a remediation (if applicable) for asynchronous actions like clicking on an email magic link or okta verify.
```

**Review by JayNewstrom — COMMENTED** [review]
Looks good. A few more small comments.

**Comment by JayNewstrom** [comment]
Once you address the last few comments, could you rebase your fork?

**Review by JayNewstrom — APPROVED** [review]

**Comment by JayNewstrom** [comment]
> Ah sorry I didn't see your last comment and merged master.

That's okay. I can squash and merge.

**Comment by deepu105** [comment]
awesome

Thanks & Regards,
Deepu

**Pull request merged**